app/src/pages/playground/PlaygroundOutput.tsx

```diff
@@ -67,80 +104,367 @@ function PlaygroundOutputMessage({ message }: { message: ChatMessage }) {
+    params: chatCompletionParams,
+    onNext,
+    onCompleted: () => {
+      markPlaygroundInstanceComplete(props.playgroundInstanceId);
+    },
+    onFailed: (error) => {
+      // TODO(apowell): We should display this error to the user after formatting it nicely.
+      // eslint-disable-next-line no-console
+      console.error(error);
+      markPlaygroundInstanceComplete(props.playgroundInstanceId);
+      updateInstance({
+        instanceId: props.playgroundInstanceId,
+        patch: {
+          activeRunId: null,
+        },
+      });
+      notifyError({
+        title: "Failed to get output",
+        message: "Please try again.",
+      });
+    },
+  });
 
-      return messages.map((message, index) => {
-        return <PlaygroundOutputMessage key={index} message={message} />;
+  const onCompleted = useCallback(
+    (
+      response: PlaygroundOutputMutation$data,
+      errors: PayloadError[] | null
+    ) => {
+      markPlaygroundInstanceComplete(props.playgroundInstanceId);
+      updateInstance({
+        instanceId,
+        patch: {
+          spanId: response.chatCompletion.span.id,
+        },
       });
+      if (errors) {
+        notifyError({
+          title: "Chat completion failed",
+          message: errors[0].message,
+        });
+        return;
+      }
+      if (response.chatCompletion.errorMessage != null) {
+        notifyError({
+          title: "Chat completion failed",
+          message: response.chatCompletion.errorMessage,
+        });
+        return;
+      }
+      setOutputContent(response.chatCompletion.content ?? undefined);
+      setToolCalls(response.chatCompletion.toolCalls);
+    },
+    [
+      instanceId,
+      markPlaygroundInstanceComplete,
+      notifyError,
+      props.playgroundInstanceId,
+      updateInstance,
+    ]
+  );
+
+  useEffect(() => {
+    if (!hasRunId) {
+      return;
     }
-    if (typeof instance.output === "string") {
-      return (
-        <PlaygroundOutputMessage
-          message={{
-            id: generateMessageId(),
-            content: instance.output,
-            role: "ai",
-          }}
-        />
-      );
+    setOutputContent(undefined);
+    setToolCalls([]);
+
+    if (streaming) {
+      startStreaming();
+      return;
     }
-    return "click run to see output";
-  }, [hasRunId, instance.output, instanceId, runId]);
+    generateChatCompletion({
+      variables: {
+        input: chatCompletionParams,
+      },
+      onCompleted,
+      onError(error) {
+        markPlaygroundInstanceComplete(props.playgroundInstanceId);
+        notifyError({
+          title: "Failed to get output",
+          message: error.message,
+        });
+      },
+    });
+    // Remove startSubscription from dependencies as its reference is not stable and we don't want to restart the subscription when it changes
+    // eslint-disable-next-line react-compiler/react-compiler
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [
```

Comment:
it might be better to remove more things from this deps array, but for now i was trying to keep it as exhaustive as possible to avoid any stale values, the startSubscription causes an infinite loop so it has to be removed.